### Lecture 3 -- Code Optimisation & Profiling

---

#### Milestone 1 -- Function-level profiling

In [16]:
import cProfile
import pstats
from mandelbrot_naive import compute_mandelbrot_naive
from mandelbrot_numpy import compute_mandelbrot_numpy

cProfile.run(
    "compute_mandelbrot_naive(100,1024,1024)",
    "naive_profile.prof"
)

cProfile.run(
    "compute_mandelbrot_numpy(100,1024,1024)",
    "numpy_profile.prof"
)

for name in ("naive_profile.prof", "numpy_profile.prof"):
    stats = pstats.Stats(name)
    stats.sort_stats("cumulative")
    stats.print_stats(10)

Tue Mar  3 17:25:34 2026    naive_profile.prof

         23067815 function calls in 19.834 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   19.838   19.838 {built-in method builtins.exec}
        1    0.004    0.004   19.837   19.837 <string>:1(<module>)
        1    2.141    2.141   19.834   19.834 c:\Users\braec\Desktop\8.th semester\Numerical scientific computing\mandelbrot-nsc\MiniProject1\mandelbrot_naive.py:21(compute_mandelbrot_naive)
  1048576   13.946    0.000   17.557    0.000 c:\Users\braec\Desktop\8.th semester\Numerical scientific computing\mandelbrot-nsc\MiniProject1\mandelbrot_naive.py:11(mandelbrot_point)
 22018211    3.611    0.000    3.611    0.000 {built-in method builtins.abs}
     1024    0.132    0.000    0.132    0.000 c:\Users\braec\Desktop\8.th semester\Numerical scientific computing\mandelbrot-nsc\MiniProject1\mandelbrot_naive.py:23(<listcomp>)
        1    0.000    0

### Questions to Answer

---

#### Which function takes the most total time?

The **naive implementation** is slowest, with a runtime of **19.834 seconds**.  
most of the time is spent in `compute_mandelbrot_naive`, which is dominated by repeated calls to lower-level helper functions.

---

#### Are there functions called surprisingly many times?

- `mandelbrot_point` is called **1,048,576 times**, which is expected since it is executed once per pixel.
- However, the function `abs` is called **22,018,211 times**, which is surprisingly high.  
  Corresponding to an average of **approximately 21 calls per pixel**, making it a major contributor to the overall runtime.

---

#### How does the NumPy profile compare to the naive version?

The NumPy implementation performs only **377 function calls**, compared to **23,067,815** calls in the naive version.  
This drastic reduction in Python-level function calls results in a runtime of approximately **2 seconds**, yielding a **~10× speedup** over the naive implementation.

Profiling shows that the naive version is dominated by Python-level loops and repeated calls to `mandelbrot_point` and `abs`, whereas the NumPy version spends most of its time in optimized, vectorized C routines.